### PR TITLE
Adding helper functions to HListT

### PR DESCRIPTION
**Pull request by JackMatusiewicz**
Adding a couple of functions that allow you to take an HListT and return either the HList or the list components. Also added some tests to confirm it works as expected.

### CODE REVIEW

File: HCollections.Test/TestHListT.fs [hunk 1]
```diff
@@ -35,3 +34,42 @@ module TestHListT =
...
+        Assert.Equal<int list> ([10; 4], testHlist)
+
+    [<Fact>]
+    let ``HListT.toList on an empty HListT returns an empty list`` () =
+        let testList = HListT.toList HListT.empty<int>
+
+        Assert.Equal<int list> ([], testList)
+
+    [<Fact>]
+    let ``HListT.toHList returns the correct HList type`` () =
```
Comment (JackMatusiewicz) [inline HCollections.Test/TestHListT.fs:55]: Not a huge fan of how I'm testing the HList component. However, in order to do it properly I'd need to add an equality function to the library (or maybe the DU internal rather than private and then write something in the tests). Happy to do that if you agree.

Comment (Smaug123) [reply HCollections.Test/TestHListT.fs:55]: I don't really want an equality function in the library, because then the type constraints would become nightmarish (what if you want an HList whose types are constrained only to be equatable? we should only do that after some thought). `toTypes` is absolutely fine.

File: HCollections/HListT.fs [hunk 2]
```diff
@@ -71,6 +71,45 @@ module HListT =
                         xs |> Teq.castFrom teq
                 }
 
+    let rec private toHList'<'ts, 'elem, 'k>
+        (input : HListT<'ts, 'elem>)
+        (cont : 'ts HList -> 'k)
+        : 'k
+        =
+        match input with
```
Comment (Smaug123) [inline HCollections/HListT.fs:79]: Is it really important to avoid stack overflows when VS can't even cope? :P

File: HCollections/HListT.fs [hunk 3]
```diff
@@ -71,6 +71,45 @@ module HListT =
                         xs |> Teq.castFrom teq
                 }
 
+    let rec private toHList'<'ts, 'elem, 'k>
+        (input : HListT<'ts, 'elem>)
```
Comment (Smaug123) [inline HCollections/HListT.fs:75]: I'm now wondering if a `HListT` should be implemented as an `'elem list` and a `'ts HList` internally. We'd get these for free then. We already have a `TypeList` internally; why should we not use an `HList` internally?

File: HCollections/HListT.fsi [hunk 4]
```diff
@@ -61,3 +61,9 @@ module HListT =
 
     /// Obtain the corresponding TypeList for the heterogeneous types in an HListT.
     val toTypeList<'ts, 'elem> : HListT<'ts, 'elem> -> TypeList<'ts>
+
+    /// Given an HListT, returns the HList component
```
Comment (Smaug123) [inline HCollections/HListT.fsi]: This description is a little hard to parse from the perspective of someone who doesn't know the library. This is morally `HListT.map snd`. I'd phrase the comment as "Given an HListT, returns the corresponding HList of heterogeneous types, discarding the homogeneous element of each entry."

Similarly below.

**Review by Smaug123 — COMMENTED** [review]
I think (cribbed from Nick really) a neater way to do this is to carry around the HList internally in an HListT. Then we can probably get away with an HListT being simply a pair of `'elem list` and `'ts HList`, and our constructor functions are the guarantee that those lists are the same length. (The public API is right already.) Would that work?

File: HCollections.Test/TestHListT.fs (hunk 1 above)
Comment (JackMatusiewicz) [reply HCollections.Test/TestHListT.fs:55]: Yeah, my personal preference would have been to create a custom equality checker in the test file. At the moment the test would pass if I just supplied Unchecked.defaultof but making massive changes isn't a price I wanted to pay unless we were in agreement.

File: HCollections/HListT.fs (hunk 3 above)
Comment (JackMatusiewicz) [reply HCollections/HListT.fs]: Can make the change

File: HCollections/HListT.fs (hunk 2 above)
Comment (JackMatusiewicz) [reply HCollections/HListT.fs:79]: My thoughts behind this was that you have no idea what the state of the stack is when you call it. So you could be really close to filling it.

File: HCollections.Test/TestHListT.fs (hunk 1 above)
Comment (nickcowle) [reply HCollections.Test/TestHListT.fs:55]: Yes, I agree that trying to define equality would be a bit of a nightmare - these collections have `[<NoComparison>]` and `[<NoEquality>]` attributes for a reason! 😄

File: HCollections/HListT.fs (hunk 3 above)
Comment (nickcowle) [reply HCollections/HListT.fs:75]: Yes, might be worth trying this out to see how it looks. You'll lose some safety in the implementation (as now we are using the API only to guarantee that the `'elem list` and the `'ts HList` are the same length, but you'll gain performance wins for when you call `toHList` and `toList`

**Review by nickcowle — APPROVED** [review]

File: HCollections/HListT.fs (hunk 3 above)
Comment (nickcowle) [reply HCollections/HListT.fs:75]: N.B if you do want to try changing the implementation, I would suggest getting this merged first and then doing it as another PR.

Comment (JackMatusiewicz) [reply HCollections/HListT.fs:75]: Sounds good to me, here's the updated HListT: https://github.com/JackMatusiewicz/HCollections/tree/HListTStoresHList

Once this has been merged I can fix the new branch up.
